codec/gcm.go

```diff
@@ -29,40 +63,59 @@ func NewGCMSession(key [32]byte) (*GCMSession, error) {
 	if err != nil {
 		return &GCMSession{}, fmt.Errorf("creating gcm cipher: %v", err)
 	}
-	return &GCMSession{
-		gcm:       gcm,
-		wRand:     rand.New(rand.NewSource(int64(binary.BigEndian.Uint64(key[:8])))),
-		wBuf:      make([]byte, gcm.NonceSize()),
-		nonceSize: gcm.NonceSize(),
-	}, nil
+
+	gcmSession := &GCMSession{
+		gcm:        gcm,
+		readNonce:  gcmNonce{},
+		writeNonce: gcmNonce{},
+	}
+
+	if bytes.Compare(self[:], remote[:]) < 0 {
+		gcmSession.writeNonce.top = math.MaxUint32
+		gcmSession.writeNonce.bottom = math.MaxUint64
+		gcmSession.writeNonce.countDown = true
+	} else {
+		gcmSession.readNonce.top = math.MaxUint32
+		gcmSession.readNonce.bottom = math.MaxUint64
+		gcmSession.readNonce.countDown = true
+	}
+	return gcmSession, nil
 }
 
+// GCMEncoder accepts a GCMSession and an encoder that wraps data encryption
 func GCMEncoder(session *GCMSession, enc Encoder) Encoder {
 	return func(w io.Writer, buf []byte) (int, error) {
-		_, err := session.wRand.Read(session.wBuf)
+		nonceBuf := [12]byte{}
+		binary.BigEndian.PutUint32(nonceBuf[:4], session.writeNonce.top)
+		binary.BigEndian.PutUint64(nonceBuf[4:], session.writeNonce.bottom)
+		session.writeNonce.next()
+		encoded := session.gcm.Seal(nil, nonceBuf[:], buf, nil)
+		n, err := enc(w, encoded)
 		if err != nil {
-			return 0, fmt.Errorf("generating randomness: %v", err)
+			return n, fmt.Errorf("Encoded sealed data: %v", err)
 		}
-		n, err := enc(w, append(session.wBuf, session.gcm.Seal(nil, session.wBuf, buf, nil)...))
-		if err != nil {
-			return n, fmt.Errorf("encoding sealed data: %v", err)
```

Comment:
Should this error also return `len(buf)` rather than n?

Reply:
I think returning zero is the only consistent option, where in the context of this encoder, the returned values is "how many decrypted/authenticated bytes where written". In the case of an error, that is typically zero.